### PR TITLE
Object cache for CBA_fnc_getObjectConfig

### PR DESCRIPTION
**Pull request by kuloodporny**
**When merged this pull request will:**
- Make CBA_fnc_getObjectConfig cache config class pointer as object variable "CBA_config"

### CODE REVIEW

**Comment by commy2** [comment]
Benchmarks please.

**Comment by dedmen** [comment]
I created a internal ticket for https://sqf.bisimulations.com/display/SQF/configOf which can probably just replace this then

**Comment by kuloodporny** [comment]
CBA v3.15.1.200418
```Result:
0.0086 ms

Cycles:
10000/10000

Code:
[player] call cba_fnc_getObjectConfig```
CBA with Artifact from this task:
```Result:
0.0062 ms

Cycles:
10000/10000

Code:
[player] call cba_fnc_getObjectConfig```

**Comment by commy2** [comment]
Now do
```sqf
typeOf player call cba_fnc_getObjectConfig
```

File: addons/common/fnc_getObjectConfig.sqf [hunk 1]
```diff
@@ -19,24 +20,39 @@ Example:
...
+if (isClass _result) exitWith {_result};
+
+// Convert object to classname to be found if needed
+private _className = _object;
+if _isObject then {
+    _className = typeOf _object;
+};
+
+// Invalid class
+if (count _className == 0) exitWith {_result};
```
Comment (PabstMirror) [inline addons/common/fnc_getObjectConfig.sqf:45]:
```suggestion
if (_className == "") exitWith {_result};
```

Comment (kuloodporny) [reply addons/common/fnc_getObjectConfig.sqf]: I used ` if (_className isEqualTo "") exitWith {_result};` as it seems to have tiny performance advantage even in this case.
```
Result:
3.69373 ms

Cycles:
271/10000

Code:
for "_i" from 0 to 10000 do {"" isEqualTo ""};
```
vs
```
Result:
3.7594 ms

Cycles:
266/10000

Code:
for "_i" from 0 to 10000 do {"" == ""};
```
I did about 20 measurements and these are the best ones I obtained, with `==` consistently slower than `isEqualTo`.

**Comment by kuloodporny** [comment]
In case of string parameter, the changed function is significantly **slower**:
Steam CBA_A3:
```
Result:
0.0074 ms

Cycles:
10000/10000

Code:
typeOf player call cba_fnc_getObjectConfig
```
This branch :
```
Result:
0.0103 ms

Cycles:
10000/10000

Code:
typeOf player call cba_fnc_getObjectConfig
```

**Comment by kuloodporny** [comment]
I have rewritten the solution a bit to make the script simpler, the results in console tests are promising - below 80 ms for text and below 50 ms for object, but I need to check it with new artifact.

**Comment by kuloodporny** [comment]
and indeed, the code with less changes is not that much worse:
```
Result:
0.0079 ms

Cycles:
10000/10000

Code:
typeOf player call cba_fnc_getObjectConfig
```
and object parameter is significantly faster:
```
Result:
0.005 ms

Cycles:
10000/10000

Code:
player call cba_fnc_getObjectConfig
```
I did over 30 tests each, the fastest ones are presented and the average results are not significantly worse.

**Comment by commy2** [comment]
Now let's benchmark different simulation types:
```sqf
mine1 = createMine ["APERSMine", ASLToAGL getPosASL player, [], 0];
```
once, then code performance:
```
mine1 call cba_fnc_getObjectConfig
```

**Comment by kuloodporny** [comment]
**Getting config of `mine1` is significantly slower with the caching enabled - over 50% overhead is present.**
Caching on:
```
Result:
0.0157 ms

Cycles:
10000/10000

Code:
mine1 call cba_fnc_getObjectConfig
```
Caching off:
```
Result:
0.01 ms

Cycles:
10000/10000

Code:
mine1 call cba_fnc_getObjectConfig
```

**Getting config for `typeOf mine1` is pretty much the same.**
Caching on:
```
Result:
0.0098 ms

Cycles:
10000/10000

Code:
typeOf mine1 call cba_fnc_getObjectConfig```
Caching off:
```
Result:
0.0097 ms

Cycles:
10000/10000

Code:
typeOf mine1 call cba_fnc_getObjectConfig```

**Comment by commy2** [comment]
It seems that the change gives an advantage for some object classes, but disfavours classnames and a few object classes. It is not clear to me that it is an all around improvement to have this cached.

**Comment by kuloodporny** [comment]
Agree, I'd say let's scratch it and wait for configOf to be implemented.

**Pull request closed without merging**